pkg/cluster/k8sres.go

```diff
@@ -1905,12 +1905,33 @@ func (c *Cluster) generateServiceAnnotations(role PostgresRole, spec *acidv1.Pos
 	for k, v := range c.OpConfig.CustomServiceAnnotations {
 		annotations[k] = v
 	}
-	if spec != nil || spec.ServiceAnnotations != nil {
-		for k, v := range spec.ServiceAnnotations {
-			annotations[k] = v
+
+	var specServiceAnnotations map[string]string
+	if spec != nil {
+		switch role {
+		case Master:
+			// MasterServiceAnnotations take precedence over ServiceAnnotations if they are not empty
+			if len(spec.MasterServiceAnnotations) > 0 {
+				specServiceAnnotations = spec.MasterServiceAnnotations
+			} else {
+				specServiceAnnotations = spec.ServiceAnnotations
+			}
+		case Replica:
+			// ReplicaServiceAnnotations take precedence over ServiceAnnotations if they are not empty
+			if len(spec.ReplicaServiceAnnotations) > 0 {
+				specServiceAnnotations = spec.ReplicaServiceAnnotations
+			} else {
+				specServiceAnnotations = spec.ServiceAnnotations
+			}
+		default:
+			specServiceAnnotations = spec.ServiceAnnotations
```

Comment:
This is either master|replica service annotation OR service annotation, right?
It should be accumulative. First, global config, then ServiceAnnotations overriding, then MasterServiceAnnotations and ReplicaServiceAnnotations. (maybe there's a golang function so we don't need 4 loops)

Reply:
Addressed in https://github.com/zalando/postgres-operator/pull/2161/commits/03878300fb17c5c46e7b621551ba8b249a4aca2e